src/graphapi/reshape.cpp

```diff
@@ -0,0 +1,166 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/errors.hpp>
+#include <miopen/graphapi/reshape.hpp>
+
+namespace miopen {
+
+namespace graphapi {
+
+const std::string& OperationReshape::signName() const
+{
+    static const std::string name = "OP_RESHAPE";
+    return name;
+}
+
+std::vector<Tensor*> OperationReshape::getInTensors() const { return {mX}; }
+
+std::vector<Tensor*> OperationReshape::getOutTensors() const { return {mY}; }
+
+OperationReshapeBuilder& OperationReshapeBuilder::setX(Tensor* x)
+{
+    mOperationReshape.mX = checkPtr(x);
+    return *this;
+}
+
+OperationReshapeBuilder& OperationReshapeBuilder::setY(Tensor* y)
+{
+    mOperationReshape.mY = checkPtr(y);
+    return *this;
+}
+
+OperationReshape OperationReshapeBuilder::build()
+{
+    if(mOperationReshape.mX != nullptr && mOperationReshape.mY != nullptr &&
```

Comment:
Do you have checks somewhere regarding sizes of input and output tensor? For example if input tensor is "n x h x 2 x 3"
then output tensor should be "n x h x 3 x 2".

@CAHEK7  @amberhassaan what do you think guys. Would be nice to have, right ?

Reply:
yes, as long as we only support `transpose` in `reshape`.

Reply:
I was thinking about the condition here. Are we supporting a transpose only? Are we restricting it to 4-dimensional tensors? And are we restricting it for the last 2 dimensions?

Reply:
In MatMul nide,  you guys, restricted minimum number of dimensions to 2, which probably makes sense. 
So may do in a pretty same way. But, yes, if you have a tensor with dims [n,h,s,d] then transposed version of this tensor would be [n,h,d,s] in my understanding

Reply:
In MIOpen we support `NCHW` to `NHWC` transpose, so that suggests that `transpose` can be more general than that. Similarly, `reshape` can do things beyond `transpose`, e.g., treat a `[n, h, w, c]` tensor as `n, h*w, c, 1` tensor. So the question then becomes: should we support `reshape` in the long run or is it a temporary thing for BWD MHA?

Reply:
I believe we can merge it and enhance later if needed.

Reply:
I'm going to restrict `reshape` node to `transpose` for the last 2 dimensions. But the check if a `reshape` node is a `transpose` node should nevertheless be repeated in graph matching/extracting because `reshape` descriptor might be extended in the future.

Reply:
@scerzh : can you please introduce an enum that specifies whether the reshape operation is `reshape` or `transpose`? Or do we guess from `tensor dims` (per cuDNN)?

Reply:
The user provides only dims so we should guess from them. But still we can add an enum property which in any case will be calculated during `finalize` and also it will avoid code duplication. I think it's a good idea.